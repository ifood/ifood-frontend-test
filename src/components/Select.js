import { useState, useEffect } from 'react'
import { array, func, string } from 'prop-types'
import styled from 'styled-components'

const Select = ({ id, onChange, options, text }) => {
  const [valueSelected, setValueSelected] = useState('')

  useEffect(() => {
    if (valueSelected) {
      onChange(id, valueSelected)
    }
  }, [id, onChange, valueSelected])

  return (
    <SelectStyle
      className="wrapper-filter-input wrapper-filter-select"
      valueSelected={valueSelected}
    >
      {options && options.length ? (
        <>
          <select
            className="filter-input filter-select"
            onChange={(event) => setValueSelected(event.target.value)}
          >
            <option />
            {options.map((option) => (
              <option key={option.value}>{option.name}</option>
            ))}
          </select>
          <label id="label-select">{text}</label>
        </>
      ) : null}
    </SelectStyle>
  )
}

const valueActive = `
  select.filter-select {
    color: white;
  }

  #label-select {
    color: white;
    top: -10px;
    z-index: 1;
  }
`

const SelectStyle = styled.div`
  .filter-select {
    cursor: pointer;
    -webkit-appearance: none;
    -moz-appearance: none;

    &:focus + label {
      color: white;
      top: -10px;
      z-index: 1;
    }
  }

  ${({ valueSelected }) => (valueSelected ? valueActive : '')}
`

Select.propTypes = {
  id: string.isRequired,
  onChange: func.isRequired,
  options: array.isRequired,
  text: string.isRequired
}

export default Select
