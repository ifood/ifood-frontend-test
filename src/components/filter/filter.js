import React, { Component } from 'react';
import TextField from '@material-ui/core/TextField';
import RenderField from '../render-field/render-field';
import axios from 'axios';
import {URL_GET_FILTER, MOCKY_API_ERROR_MESSAGE} from '../../constants';

class Filter extends Component {
    constructor(props) {
      super(props);
  
      this.state = {
        filtersFields: []
      };
    }

    // Method called before render the page
    componentDidMount() {
      axios.get(URL_GET_FILTER) // Call mocky API to get the filter fields/Values
        .then(response => {
            this.setState({
            filtersFields: response.data.filters
            })
        })
        .catch(error => {
            console.log(MOCKY_API_ERROR_MESSAGE, error)
        });
    }

    // Method to render the page
    render() {
      const { filtersFields } = this.state;
      const {
        filterValues,
        onChangeFilters,
      } = this.props;

      return (
        <div>
          {
            filtersFields ?
              <div style={{ display: 'flex', flexDirection: 'column' }}>
                <TextField
                  id="search"
                  label="Filter playlists by name"
                  onKeyUp={event => this.props.onTextChange(event.target.value)}
                  style={{ marginBottom: '.8em' }} />
                {
                  filtersFields.map(field => ((
                    <RenderField
                      {...field}
                      key={field.id}
                      onChange={onChangeFilters}
                      fieldValues={filterValues}
                    />
                  )))
                }
              </div> : 'Error'
          }
        </div>
      );
    }
  }

  export default Filter;