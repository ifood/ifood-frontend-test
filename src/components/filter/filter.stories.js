import React from 'react';

import Filter from '.';

import { playlistResponseMocks } from '../../services/utils/mocks/playlist.response';

export default {
    title: 'Components/Filter',
    component: Filter,
    argTypes: {
      parameters: '?',
      playlists: playlistResponseMocks
    },
};

const Template = (args) => <Filter {...args} />;

export const Default = Template.bind({});
Default.args = {
  parameters: '?',
  playlists: playlistResponseMocks
};