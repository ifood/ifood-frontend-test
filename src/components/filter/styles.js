import styled from 'styled-components';
import { desktop, mobile } from '../../services/utils/styles/media-queries';

export const Menu = styled.div`
  margin-top: 60px;
`;

export const Search = styled.input`
  border-radius: 20px;
  margin: 10px 0;
  outline: none;
  padding: 5px;
  position: fixed;
  text-align: center;
  top: 10px;
  width: calc(100vw - 30px);
`;

export const Filters = styled.section`
  align-items: center;
  border-bottom: 1px solid #fff;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 10px;

  ${mobile} {
    align-items: normal;
    flex-direction: column;
  }

  ${desktop} {
    flex-direction: row;
  }
`;

export const Filter = styled.div`
`;

export const Name = styled.p`
  color: #fff;
  margin: 0 0 5px;
`;

export const ResetButton = styled.button`
  cursor: pointer;
  border-radius: 20px;
  bottom: 40px;
  padding: 15px 10px;
  position: absolute;
`;
