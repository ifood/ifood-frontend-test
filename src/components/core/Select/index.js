import React, { useState } from "react";
import PropTypes from "prop-types";
import { useFormContext } from "react-hook-form";

import * as S from "./styles";

const Select = ({ register, optionId, values }) => {
  const [isOpen, setIsOpen] = useState(false);
  const { setValue } = useFormContext();

  return (
    <>
      <S.FieldSelect
        ref={register}
        id={optionId}
        name={optionId}
        type="text"
        onClick={() => setIsOpen(true)}
        readOnly
      />
      {isOpen ? (
        <>
          <S.Close onClick={() => setIsOpen(false)}>x</S.Close>
          <S.Options>
            {values?.map(({ name, value }) => (
              <S.Option
                key={value}
                onClick={() => {
                  setValue(optionId, value);
                  setIsOpen(false);
                }}
              >
                {name}
              </S.Option>
            ))}
          </S.Options>
        </>
      ) : (
        <S.Arrow onClick={() => setIsOpen(true)} />
      )}
    </>
  );
};

Select.propTypes = {
  register: PropTypes.func.isRequired,
  optionId: PropTypes.string.isRequired,
  values: PropTypes.array.isRequired,
};

export default Select;
