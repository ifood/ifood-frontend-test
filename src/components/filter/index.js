import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import * as FilterService from '../../services/filter';

import * as S from './styles';

import Select from './select';
import Input from './input';

function Filter({ parameters, playlists, setFilteredPlaylist, setParameters }) {
    const [filters, setFilters] = useState([]);

    useEffect(() => {
        const _getDataFilter = async () => {
          const response = await FilterService.getFilters();
          setFilters(response.filters);
        }

        _getDataFilter();
    }, []);

    const _findPlaylist = (termToFind) => {
        const foundItem = playlists.filter((playlist) =>
            playlist.name.toLowerCase().includes(termToFind.toLowerCase())
        );
        setFilteredPlaylist(foundItem);
    }

    const _handleChange = (event) => {
        _findPlaylist(event.target.value);
    };

    const mountParam = (event) => {
        let property = event.target.id;
        let value = event.target.value;

        if (property === 'timestamp') value = value.concat(":00");
        
        if (parameters.includes(property)) {
            return setParameters(parameters.replace(/[a-z]*=(\w+)/g, `${property}=${value}`));
        }

        setParameters(`${parameters}&${property}=${value}`);
    };

    const resetParameters = () => {
        setParameters(`?`);
    };

    return (
        <S.Menu>
            <S.Search 
                id="search" 
                name="search" 
                onChange={_handleChange}
                placeholder="O que vocÃª deseja ouvir?"
                type="search" 
            />

            <S.Filters>
                {filters.length > 0 && filters.map((item) => (
                    <S.Filter key={item.id}>
                        <S.Name>{item.name}</S.Name>

                        {item.values && (
                            <Select 
                                defaultValue="Selecionar"
                                id={item.id}
                                mountParam={mountParam}
                                values={item.values}
                            />
                        )}

                        {item.validation && (
                            <Input
                                defaultValue={item.id !== 'timestamp' ? 1 : null}
                                id={item.id}
                                max={item.validation.max}
                                min={item.validation.min}
                                mountParam={mountParam}
                                type={item.validation.primitiveType === 'INTEGER' ? 'number' : 'datetime-local'}
                            />
                        )}
                    </S.Filter>
                ))}
            </S.Filters>
            <S.ResetButton onClick={resetParameters}>Resetar Filtros</S.ResetButton>
        </S.Menu>
    );
}

Filter.propTypes = {
    parameters: PropTypes.string.isRequired,
    playlists: PropTypes.array.isRequired,
    setFilteredPlaylist: PropTypes.func.isRequired,
    setParameters: PropTypes.func.isRequired,
};

export default Filter;
