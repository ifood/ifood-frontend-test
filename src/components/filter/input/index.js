import React from 'react';
import PropTypes from 'prop-types';

import * as S from './styles';

function Input({ defaultValue, id, max, min, mountParam, type }) {
    return (
        <S.Input 
            defaultValue={defaultValue}
            id={id}
            max={max}
            min={min}
            name={id}
            onChange={mountParam}
            type={type}
        />
    );
}

Input.propTypes = {
    defaultValue: PropTypes.number,
    id: PropTypes.string.isRequired,
    max: PropTypes.number,
    min: PropTypes.number,
    mountParam: PropTypes.func.isRequired,
    type: PropTypes.string.isRequired,
};

Input.defaultProps = {
    min: 0,
};

export default Input;