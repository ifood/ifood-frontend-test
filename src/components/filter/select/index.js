import React from 'react';
import PropTypes from 'prop-types';

import * as S from './styles';

function Select({ defaultValue, id, mountParam, values }) {
    return (
        <S.Select defaultValue={defaultValue} id={id} onChange={mountParam}>
            <S.Option value={defaultValue} disabled hidden>{defaultValue}</S.Option>
            {values.map((option) => (
                <S.Option key={option.value} value={option.value}>{option.name}</S.Option>
            ))}
        </S.Select>
    );
}

Select.propTypes = {
    defaultValue: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    mountParam: PropTypes.func.isRequired,
    values: PropTypes.arrayOf(
        PropTypes.shape({
          value: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
        }),
    ).isRequired,
};

export default Select;