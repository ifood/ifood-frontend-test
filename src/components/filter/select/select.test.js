import React from 'react';
import { create, act } from 'react-test-renderer';
import Select from '.';

import { filterResponseMock } from '../../../services/utils/mocks/filter.response';

describe('Select Component', () => {
    const mockFunc = jest.fn();
    let root;
    
    beforeAll(() => {
        act(() => {
            root = create(
                <Select 
                    defaultValue='Click'
                    id='id-test'
                    mountParam={mockFunc}
                    values={filterResponseMock[0].values}
                />
            )
        });
    });

    it('should to be defined', () => {
        expect(Select).toBeDefined();
    });

    it('should to be match Snapshot', () => {
        expect(root.toJSON()).toMatchSnapshot();
    });
});
