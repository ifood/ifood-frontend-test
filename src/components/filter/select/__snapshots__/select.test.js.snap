// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component should to be match Snapshot 1`] = `
<select
  className="sc-AxjAm fDEmJb"
  defaultValue="Click"
  id="id-test"
  onChange={[MockFunction]}
>
  <option
    className="sc-AxirZ dqVvws"
    disabled={true}
    hidden={true}
    value="Click"
  >
    Click
  </option>
  <option
    className="sc-AxirZ dqVvws"
    value="en_AU"
  >
    en_AU
  </option>
  <option
    className="sc-AxirZ dqVvws"
    value="de_DE"
  >
    de_DE 
  </option>
  <option
    className="sc-AxirZ dqVvws"
    value="pt_BR"
  >
    pt_BR
  </option>
  <option
    className="sc-AxirZ dqVvws"
    value="fr_FR"
  >
    fr_FR
  </option>
  <option
    className="sc-AxirZ dqVvws"
    value="en_US"
  >
    en_US
  </option>
  <option
    className="sc-AxirZ dqVvws"
    value="es_AR"
  >
    es_AR
  </option>
</select>
`;
